xmtp_id/src/associations/signature.rs

```diff
@@ -60,181 +102,253 @@ impl Clone for Box<dyn Signature> {
-    signature_text: String,
-    signature_bytes: Vec<u8>,
+    message: String,
+    public_key: [u8; 64],
+    bytes: [u8; 65],
 }
 
 impl RecoverableEcdsaSignature {
-    pub fn new(signature_text: String, signature_bytes: Vec<u8>) -> Self {
-        RecoverableEcdsaSignature {
-            signature_text,
-            signature_bytes,
-        }
+    pub fn new(
+        message: String,
+        public_key: [u8; 64],
+        bytes: [u8; 65],
+    ) -> Result<Self, SignatureError> {
+        let sig = RecoverableEcdsaSignature {
+            message,
+            public_key,
+            bytes,
+        };
+        sig.verify()?;
+        Ok(sig)
     }
 }
 
 impl Signature for RecoverableEcdsaSignature {
-    fn recover_signer(&self) -> Result<MemberIdentifier, SignatureError> {
-        todo!()
+    fn get_kind(&self) -> &SignatureKind {
+        &SignatureKind::Erc191
     }
 
-    fn signature_kind(&self) -> SignatureKind {
-        SignatureKind::Erc191
+    fn get_signer(&self) -> MemberIdentifier {
+        MemberIdentifier::Address(h160addr_to_string(public_key_to_address(self.public_key)))
     }
 
-    fn bytes(&self) -> Vec<u8> {
-        self.signature_bytes.clone()
+    fn to_bytes(&self) -> Vec<u8> {
+        self.bytes.to_vec()
     }
 
     fn to_proto(&self) -> SignatureProto {
         SignatureProto {
             signature: Some(SignatureKindProto::Erc191(RecoverableEcdsaSignatureProto {
-                bytes: self.bytes(),
+                message: self.message.clone(),
+                public_key: self.public_key.to_vec(),
+                bytes: self.to_bytes(),
             })),
         }
     }
+
+    fn verify(&self) -> Result<(), SignatureError> {
+        let signature = ethers::types::Signature::try_from(self.to_bytes().as_slice())?;
+        signature.verify(self.message.clone(), public_key_to_address(self.public_key))?;
+        Ok(())
+    }
+}
+
+/// Convert a k256 public key to an ethereum [Address]
+pub fn public_key_to_address(public_key: [u8; 64]) -> Address {
+    let hash = keccak256(&public_key[..]);
+    Address::from_slice(&hash[12..])
 }
 
-#[allow(dead_code)]
 #[derive(Clone)]
 pub struct Erc1271Signature {
-    signature_text: String,
-    signature_bytes: Vec<u8>,
+    message: String,
+    bytes: Vec<u8>,
+    chain: u64,
     contract_address: String,
     block_number: u64,
 }
 
 impl Erc1271Signature {
     pub fn new(
-        signature_text: String,
-        signature_bytes: Vec<u8>,
+        message: String,
+        bytes: Vec<u8>,
+        chain: u64,
         contract_address: String,
         block_number: u64,
-    ) -> Self {
-        Erc1271Signature {
-            signature_text,
-            signature_bytes,
+    ) -> Result<Self, SignatureError> {
+        let sig = Erc1271Signature {
+            message,
+            bytes,
+            chain,
             contract_address,
             block_number,
-        }
+        };
+        sig.verify()?;
```

Comment:
I do like this change to doing all the verification as part of `new`. Maybe we want to rename the types to `ValidatedSignature` or something to make it more obvious to callers that these are safe.

Reply:
I was thinking about this a bit more. I'm not sure verifying the signatures as part of `new()` is a good idea here.

- This change puts the validation logic in two places. Some of it happens during `apply_update` and some of it happens during deserialization. That makes errors harder to handle and manage. For example, if we want to allow some Identity Updates verification failures to just cause the update to be skipped instead of aborting the entire process, we would have to do that in two places.
- It means it's not possible to deserialize an identity update without also verifying the signatures. That should probably be optional and up to the developer.
- We're now doing network calls to verify 1271 signatures inside `TryFrom` statements, which is pretty unintuitive. We have to spin up a new Runtime to do async work inside a synchronous process, which we'll have to be super careful about since it can lead to deadlocks or weird performance characteristics.

Maybe a better pattern is to have two signature structs. An `UnverifiedSignature` and a `VerifiedSignature`. Converting from an unverified -> verified could be an async process to allow for 1271 network calls. We can then make `update_state` async and do it as part of that process. The previous `recover_signer()` method would also still work, if we made it async.

Reply:
The main reason I wanted to verify the signature during `new()` was I didn't want a signature to be verified multiple time during its lifetime since it's expensive especially for 1271.

What you said makes sense, I think I'll keep it simple by sticking to the original design that has one `Signature` trait with `recover_signer()` and have all functionalities done. I can think about `UnverifiedSignature` and `VerifiedSignature` after that.

Reply:
The good news is that the code path that actually looks at these signatures is pretty narrow, so we can be careful to only call `recover_signer` once and then re-use that value.

But you're right, we do only want to do it once.